src/main/java/com/offz/spigot/staminaclimb/stamina/StaminaTask.kt

```diff
@@ -6,53 +6,107 @@ import com.offz.spigot.staminaclimb.climbing.ClimbBehaviour
 import com.offz.spigot.staminaclimb.config.StaminaConfig
 import org.bukkit.Bukkit
 import org.bukkit.GameMode
+import org.bukkit.util.Vector
+import org.bukkit.Location
+import org.bukkit.entity.Player
 import org.bukkit.boss.BarColor
 import org.bukkit.potion.PotionEffect
 import org.bukkit.potion.PotionEffectType
 import org.bukkit.scheduler.BukkitRunnable
+import java.util.Calendar
+import java.util.UUID
+import kotlin.math.abs
 
 class StaminaTask : BukkitRunnable() {
+
+    private var previousTickTime: Long = 0;
+    private val targetDeltaTime = 1.0 / 20.0
+
+    // Not sure how to get the World to make a Location so I'm just doing this | TODO: change to a Location
+    private var prevX = 0.0
+    private var prevY = 0.0
+    private var prevZ = 0.0
+
+    private var ticksSinceLastColorSwitch = 0
+    private val ticksToSwitchColor = 5
+
+    private var currentWallMultiplier = 1.0
+
+    internal fun isPlayerMoving(player: Player): Boolean {
+        val minMovementValue = 0.007
+        val isMoving = abs(player.getLocation().getX() - prevX) >= minMovementValue ||
+                       abs(player.getLocation().getY() - prevY) >= minMovementValue ||
+                       abs(player.getLocation().getZ() - prevZ) >= minMovementValue 
+        prevX = player.getLocation().getX()
+        prevY = player.getLocation().getY()
+        prevZ = player.getLocation().getZ()
+        return isMoving
+    }
+
     override fun run() {
         StaminaBar.registeredBars.keys.forEach { uuid ->
             val player = Bukkit.getPlayer(uuid) ?: StaminaBar.registeredBars.remove(uuid).let { return@forEach }
             val bar = StaminaBar.registeredBars[uuid] ?: return@forEach
             val progress = bar.progress
 
-            if (player.gameMode == GameMode.CREATIVE || player.gameMode == GameMode.SPECTATOR) {
-                StaminaBar.unregisterBar(uuid)
-                player.stopClimbing()
-                player.allowFlight = true
-                return
-            }
-
-            bar.isVisible = progress + StaminaConfig.data.staminaRegen <= 1 //hide when full
+            bar.isVisible = progress + StaminaConfig.data.staminaRegen <= StaminaConfig.data.barMax //hide when full
 
-            //regenerate stamina for BossBar
-            if (!uuid.isClimbing)
-                bar.progress = (bar.progress +
-                        //TODO this is determined by client and easily spoofable
-                        if (player.isOnGround)
-                            StaminaConfig.data.staminaRegen
-                        else StaminaConfig.data.staminaRegenInAir
-                        ).coerceAtMost(1.0)
+            if (!uuid.isClimbing) { //Regenerate stamina
+                if(player.isOnGround)
+                    StaminaBar.addProgressWithDeltaTime(StaminaConfig.data.staminaRegen
```

Comment:
In line with what I mentioned in the top comment, the auto formatter will *really* change these lines, likely for the worse unless we manually format them in a cleaner way that still follows the style guide.